finality-aleph/src/sync/mod.rs

```diff
@@ -0,0 +1,92 @@
+use std::{
+    fmt::{Debug, Display},
+    hash::Hash,
+};
+
+mod substrate;
+
+/// The identifier of a block, the least amount of knowledge we can have about a block.
+pub trait BlockIdentifier: Clone + Hash + Debug + Eq {
+    /// The block number, useful when reasoning about hopeless forks.
+    fn number(&self) -> u32;
+}
+
+/// Informs the sync that it should attempt to acquire the specified data.
+pub trait Requester<BI: BlockIdentifier> {
+    /// The sync should attempt to acquire justifications for this block.
+    fn request_justification(&self, id: BI);
+}
+
+/// The header of a block, containing information about the parent relation.
+pub trait Header: Clone {
+    type Identifier: BlockIdentifier;
+
+    /// The identifier of this block.
+    fn id(&self) -> Self::Identifier;
+
+    /// The identifier of this block's parent.
+    fn parent_id(&self) -> Option<Self::Identifier>;
+}
+
+/// The justification of a block, including a header.
+pub trait Justification: Clone {
+    type Header: Header;
+
+    /// The header of the block.
+    fn header(&self) -> &Self::Header;
+}
+
+/// A verifier of justifications.
+pub trait Verifier<RJ, J: Justification> {
+    type Error: Display;
+
+    /// Verifies the raw justification and returns a full justification if successful, otherwise an
+    /// error.
+    fn verify(&self, header: J::Header, raw_justification: RJ) -> Result<J, Self::Error>;
+}
+
+/// A facility for finalizing blocks using justifications.
+pub trait Finalizer<J: Justification> {
+    type Error: Display;
+
+    /// Finalize a block using this justification. Since the justification contains the header, we
+    /// don't need to additionally specify the block.
+    fn finalize(&self, justification: J) -> Result<(), Self::Error>;
+}
+
+/// A notification about the chain state changing.
+pub enum ChainStateNotification<BI: BlockIdentifier> {
+    /// A block has been imported.
+    BlockImported(BI),
+    /// A block has been finalized.
+    BlockFinalized(BI),
+}
+
+/// A stream of notifications about the chain state in the database changing.
+#[async_trait::async_trait]
+pub trait ChainStateNotifier<BI: BlockIdentifier> {
+    /// Returns a chain state notification when it is available.
+    async fn next(&self) -> ChainStateNotification<BI>;
+}
+
+/// The state of a block in the database.
+pub enum BlockState<J: Justification> {
```

Comment:
suggestion: Add variant `IsOnFork(J::Header)` (or some better name)

Reply:
What would this mean?

Reply:
Block is from fork that is not the best/finalized chain

Reply:
but totally optional

Reply:
I guess it could only apply to blocks that are not in finalized chain, as the ones not from best chain could still be in the best chain in the future.

This might be useful to know that a specific block will never be finalized. But is there a case where it could be useful to us?

Reply:
We will be internally figuring out which blocks are hopeless forks anyway, and we might want to have our own best-block logic. In general we want to treat the DB more like a bag of blocks and what it thinks about their structure beyond "there is a finalized blockchain" is mostly irrelevant.